src/domain/ln-invoice.ts

```diff
@@ -0,0 +1,43 @@
+import { ok, err, Result } from "neverthrow"
+import lightningPayReq from "bolt11"
+import { toTypedError } from "./utils"
+
+const toLnInvoiceDecodeError = toTypedError<LnInvoiceDecodeErrorType>({
+  unknownMessage: "Unknown decoding error",
+  _type: "LnInoviceDecodeError",
+})
+
+const safeDecode = Result.fromThrowable(lightningPayReq.decode, toLnInvoiceDecodeError)
+
+export const decodeInvoice = (
+  bolt11EncodedInvoice: string,
+): Result<LnInvoice, LnInvoiceDecodeError> => {
+  let paymentHash: PaymentHash | null = null
+  let paymentSecret: PaymentSecret | null = null
+  const paymentRequest: EncodedPaymentRequest = {
+    inner: bolt11EncodedInvoice,
+  }
+
+  return safeDecode(bolt11EncodedInvoice).andThen(({ tags }) => {
+    tags.forEach((tag) => {
+      if (tag.tagName === "payment_hash") {
```

Comment:
For similar cases like this, could an option be to use a `switch` statement with error returns to remove the "appearance" of nesting?

Example:
```ts
tags.forEach((tag) => {
  const tagError = typeof tag.data != "string"
  switch (tag.tagName) {
    case "payment_hash":
      if (tagError) {
        return err(toLnInvoiceDecodeError({ message: "Irregular payment_hash" }))
      }
      paymentHash = { inner: tag.data }
      break

    case "payment_secret":
      if (tagError) {
        return err(toLnInvoiceDecodeError({ message: "Irregular payment_secret" }))
      }
      paymentSecret = { inner: tag.data }
      break
  }
  if (paymentHash == null || paymentSecret == null) {
    ...
```

Reply:
Yup that would be cleaner. Feel free to add a commit